### PR TITLE
Feature synthetic data

### PR DESCRIPTION
**Pull request by dilpath**
Ideas for tests?

To use this, the `feature_synthetic_data` branch of AMICI and PEtab (and pyPESTO for the example) should be used together.
Implementation in AMICI: [PR](https://github.com/AMICI-dev/AMICI/pull/1283)
Example optimisation with synthetic data: [notebook](https://github.com/ICB-DCM/pyPESTO/blob/feature_synthetic_data/doc/example/synthetic_data.ipynb), [PR](https://github.com/ICB-DCM/pyPESTO/pull/482)
PEtab base class: [code](https://github.com/PEtab-dev/PEtab/blob/feature_synthetic_data/petab/simulate.py), [PR](https://github.com/PEtab-dev/PEtab/pull/472)

### CODE REVIEW

**Comment by paulstapor** [comment]
> Ideas for tests?

Given a random seed, the synthetic data should reproducible, right? At least, if noise is added to the simulation, it should be the same...
Beyond that, one could store a reference trajectory for a test model, simulate, and test the residuals to the reference via a chi2 test. Would be a probabilistic test. But one could run the same simulation 100 times and see, whether or not the overall chi2-value-distribution over the runs makes sense...

File: python/amici/petab_simulate.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+from amici import AmiciModel
+from amici.petab_import import import_petab_problem
+from amici.petab_objective import simulate_petab, rdatas_to_measurement_df
+from petab.simulate import Simulator
+
+
+class PetabSimulator(Simulator):
+    """
```
Comment (dweindl) [inline python/amici/petab_simulate.py:8]: Add class description

File: python/amici/petab_simulate.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    amici_model:
+        Instance of an `amici.amici.ModelPtr` object (TODO confirm).
```
Comment (dweindl) [inline python/amici/petab_simulate.py:10]: Redundant or conflicting with typehint. "Model to simulate" ?

File: python/amici/petab_simulate.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    def _simulate_without_noise(
+            self,
+            amici_model: AmiciModel = None,
+            **kwargs
+    ):
+        """
+        See `petab.simulate.Simulator.simulate()` docstring.
```
Comment (dweindl) [inline python/amici/petab_simulate.py:18]:
```suggestion
        See :py:func:`petab.simulate.Simulator.simulate()` docstring.
```

File: python/amici/petab_simulate.py (hunk 2 above)
Comment (FFroehlich) [inline python/amici/petab_simulate.py:10]:
```suggestion
        Amici model instance.
```

**Review by dweindl — COMMENTED** [review]
The PEtab test problem here is from the test suite? The respective repo is cloned for the other PEtab test suite runs already. Would be great to use the data from there instead of adding the files here.

File: python/tests/petab_test_models/conversion_reaction/measurements.tsv [hunk 4]
```diff
@@ -0,0 +1,11 @@
+observableId	simulationConditionId	time	measurement	noiseParameters
```
Comment (FFroehlich) [inline python/tests/petab_test_models/conversion_reaction/measurements.tsv:1]: gitignore

File: python/tests/petab_test_models/conversion_reaction/observables.tsv [hunk 5]
```diff
@@ -0,0 +1,2 @@
+observableId	observableFormula	noiseFormula
```
Comment (FFroehlich) [inline python/tests/petab_test_models/conversion_reaction/observables.tsv:1]: gitignore

File: python/tests/petab_test_models/conversion_reaction/parameters.tsv [hunk 6]
```diff
@@ -0,0 +1,3 @@
+parameterId	parameterScale	lowerBound	upperBound	nominalValue	estimate
```
Comment (FFroehlich) [inline python/tests/petab_test_models/conversion_reaction/parameters.tsv:1]: gitignore

**Comment by FFroehlich** [comment]
> The PEtab test problem here is from the test suite? The respective repo is cloned for the other PEtab test suite runs already. Would be great to use the data from there instead of adding the files here.

agreed

File: python/amici/petab_simulate.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        result = simulate_petab(
+            self.petab_problem,
+            self.amici_model,
+            **{k: v
+               for k, v in kwargs.items()
+               if k in relevant_kwargs_simulate},
+        )
+
+        # TODO use `rdatas_to_simulation_df` instead?
+        simulation_df = rdatas_to_measurement_df(
```
Comment (FFroehlich) [inline python/amici/petab_simulate.py:69]:
```suggestion
        measurement_df = rdatas_to_measurement_df(
```
as this function is supposed to generate data, not simulations we should stick to `measurement_df` here.

**Comment by FFroehlich** [comment]
> > Ideas for tests?
> 
> Given a random seed, the synthetic data should reproducible, right? At least, if noise is added to the simulation, it should be the same...
> Beyond that, one could store a reference trajectory for a test model, simulate, and test the residuals to the reference via a chi2 test. Would be a probabilistic test. But one could run the same simulation 100 times and see, whether or not the overall chi2-value-distribution over the runs makes sense...

I agree this test makes sense, but I don't think it should be implemented in amici since amici doesn't implement the noise generation (should probably be done in pypesto?).

I think tests here should focus on repeated simulations and passing of kwargs to the right function.

**Comment by paulstapor** [comment]
> I agree this test makes sense, but I don't think it should be implemented in amici since amici doesn't implement the noise generation (should probably be done in pypesto?).

Wasn't aware which package does actually what in this new feature.
The test obviously makes most sense to be implemented in the package which finally generates the artificial data (and the noise)...
If this generation is done by pyPESTO, then obviously pyPESTO is the better choice, fully agreed.

**Comment by dilpath** [comment]
Thanks for the feedback. There are now tests for repeated calls without noise, and handling of `kwargs`.

Tests on noise generation have been added to the [PEtab PR](https://github.com/PEtab-dev/PEtab/pull/472).

Also switched to using the already cloned PEtab test suite (won't check this is working until the PEtab PR is merged).

**Comment by dilpath** [comment]
Now ready for review (the PEtab version in this PR is `develop`, and will be changed to the next PEtab PyPI version once the next PEtab version is released, before this PR is merged).

**Review by dweindl — APPROVED** [review]

**Comment by dweindl** [comment]
> Now ready for review (the PEtab version in this PR is `develop`, and will be changed to the next PEtab PyPI version once the next PEtab version is released, before this PR is merged).

petab 0.1.11 is now available on pypi

File: python/sdist/setup.py [hunk 8]
```diff
@@ -209,7 +209,7 @@ def main():
         setup_requires=['setuptools>=40.6.3'],
         python_requires='>=3.6',
         extras_require={
-            'petab': ['petab==0.1.7'],
+            'petab': ['petab==0.1.11'],
```
Comment (dweindl) [inline python/sdist/setup.py:212]:
```suggestion
            'petab': ['petab>=0.1.11'],
```

Would only set a minimum version requirement (assuming no major backwards incompatibilities in the near future...)

**Review by FFroehlich — APPROVED** [review]

**Pull request merged**